### PR TITLE
Mailing System (Request mentoring)

### PR DESCRIPTION
**Pull request by KaylaKwon**
Send a push mail when mentee request mentoring to mentor.
Default accout is Yoda Google accout and no authentication yet.

@ethkim @nlpenguin @SarahhhC @lyw0149

### CODE REVIEW

File: src/server/controllers/match.controller.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+/*
+ * Methods about mentoring request, accept or reject including E-mail Service
+ */
+
+let matchData = new matchSchema();
+
+let sampleSuccessResult = {
+  successCode: 1,
+  matchData: matchData,
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: what's the difference between the first matchData and the second? If they're different variables we should name them differently.

File: src/server/controllers/match.controller.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  err: null,
+};
+
+// Send mentoring request pushing Email to mentor(receiver)
+function sendRequestEmail(req, res, receiver) {
+  let transport
+    = mailer.createTransport('smtps://yoda.mentor.lab%40gmail.com:svbootcamp@!@smtp.gmail.com');
+  let mailOptions = {
+    from: '"Yoda Service Team" <yoda.mentor.lab@gmail.com>',
+    to: receiver,
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: Can we add a replyTo field to use the current user's email address?

File: src/server/controllers/match.controller.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    subject: 'New mentee needs your help!',
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: Actually, hard coded objects like this should be defined as a constant, and declared at the top (just like sampleSuccessResult).

File: src/server/controllers/match.controller.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  transport.sendMail(mailOptions, function(err, response){
+    if (err){
+      sampleFailResult.errPoint = 'RequestMentoring - transport.sendMail';
+      sampleFailResult.err = err;
+
+      res.json(sampleFailResult);
+    } else {
+      sampleSuccessResult.mailOptions = mailOptions;
+
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: So we're not actually sending an email yet? Why is that? It looks like we're almost there..

File: src/server/controllers/match.controller.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      res.json(sampleSuccessResult);
+    }
+    transport.close();
+  });
+}
+
+// The mentee sent request to Mentor
+export function requestMentoring(req, res, next) {
+  matchData.mentorId = req.body.mentorId; // Email
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: Each mentor/mentee also has user_id we defined in the user model. Can we use that? 

@SarahhhC : This is actually related to the get_mentor() method you're building, where you're extracting a subset of user object (id, email, name) and only pass that around instead of the entire user object. We have 2 options:

1) Pass around the entire user object: this makes the rest of the backend code much easier because we have the full record of the users. So we can just do the matching by 2 users (mentor/mentee). The disadvantage is that our system can get a bit bloated since you'll be returning back all the user data from db. 

2) Define the list of fields that are useful for the backend / frontend. At some point, we may decide to store the user preferences (fact that s/he likes cooking, prefers looking at mentors from good schools etc) which may not be used for the UI or any other matching service. In that case, we can decide to return back a subset of fields that we absolutely need (for example, for the "detailed view" page). If we need more fields for small set of users, we can make another db lookup to get the full set of fields. The upside of this is that the system is light. The downside is, whatever that's not included in the "user preview" object, will have to be retrieved again using find(user_id).

@KaylaKwon : Either way works fine, but the matchData should then have to store 2 objects (mentor, mentee) which is either the full user object, or a subset of user object. Let's avoid having another custom data fields like menteeId, mentorName, etc.

File: src/test/node/user.js [hunk 6]
```diff
@@ -1,9 +0,0 @@
-import test from 'tape';
```
Comment (ethkim) [inline src/test/node/user.js]: Why was this file removed?

File: src/server/models/match.model.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+import mongoose from 'mongoose';
+
+const Schema = mongoose.Schema;
+let matchSchema = new Schema({
+  mentorId: String,
```
Comment (ethkim) [inline src/server/models/match.model.js]: As discussed above, couldn't we just reuse the user object?

File: src/server/controllers/match.controller.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      sampleFailResult.errPoint = 'RequestMentoring - Saving MatchData';
+      sampleFailResult.err = err;
+
+      res.json(sampleFailResult);
+    } else {
+      // Todo:
+      // 1. update mentor and mentee's request db
+
+      // 2. Send Request Mail to Mentor
+      sendRequestEmail(req, res, 'papermoon703@gmail.com');
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: the recipient should probably come from mentorEmail?

File: src/server/controllers/match.controller.js (hunk 1 above)
Comment (KaylaKwon) [inline src/server/controllers/match.controller.js]: to change sample result dynamically according to reqest parameter,
**function requestMentoring(..)** change the content of first **matchData**.
and **sampleSuccessResult** take that values

File: src/server/controllers/match.controller.js (hunk 4 above)
Comment (KaylaKwon) [inline src/server/controllers/match.controller.js]: This code can send an email by 
`localhost:8000/match/request`
with POST method, including parameter 'mentorId', 'menteeId', 'mentorName', 'menteeName'.

mentorId and menteeId is gmail

File: src/test/node/user.js (hunk 6 above)
Comment (KaylaKwon) [inline src/test/node/user.js]: insertMockData method occurs an travis failure in un-updated code (like mine)
and will be moved to somewhere.

so I removed it

File: src/server/controllers/match.controller.js (hunk 8 above)
Comment (KaylaKwon) [inline src/server/controllers/match.controller.js]: That's right~!

File: src/server/models/match.model.js (hunk 7 above)
Comment (KaylaKwon) [inline src/server/models/match.model.js]: I was about to divide user collection and match collection.
match collection is made up with string,
but using user object seems good.

I'll reflect that

File: src/server/controllers/match.controller.js (hunk 5 above)
Comment (KaylaKwon) [inline src/server/controllers/match.controller.js]: It was from other collection named 'Match' lol

Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: I think I did option 2, maybe. I understood that list view need few info like mentor name and field. This is just 2nd sprint, so we can talk about the options~

File: src/server/controllers/match.controller.js [hunk 9]
```diff
@@ -7,99 +7,55 @@ import mailer from 'nodemailer';
...
+let matchData = new matchSchema();
 let sampleFailResult = {
   successCode: 0,
   errPoint: null,
   err: null,
 };
 
 // Send mentoring request pushing Email to mentor(receiver)
-function sendRequestEmail(req, res, receiver) {
+function sendRequestEmail(res, receiver, replyTo, content) {
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: receiver -> mentor

File: src/server/controllers/match.controller.js [hunk 10]
```diff
@@ -7,99 +7,55 @@ import mailer from 'nodemailer';
...
   let transport
     = mailer.createTransport('smtps://yoda.mentor.lab%40gmail.com:svbootcamp@!@smtp.gmail.com');
   let mailOptions = {
-    from: '"Yoda Service Team" <yoda.mentor.lab@gmail.com>',
+    from: YODA_ACCOUNT,
     to: receiver,
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: receiver -> mentor

File: src/server/models/match.model.js [hunk 11]
```diff
@@ -1,13 +1,12 @@
...
-
+// const ObjectId = mongoose.Types.ObjectId;
 const Schema = mongoose.Schema;
+
 let matchSchema = new Schema({
-  mentorId: String,
-  mentorName: String,
-  menteeId: String,
-  menteeName: String,
+  mentor: Schema.Types.ObjectId,
```
Comment (ethkim) [inline src/server/models/match.model.js]: ObjectId is an ID, right? We should call this as mentorId then.

File: src/server/models/match.model.js [hunk 12]
```diff
@@ -1,13 +1,12 @@
...
+  mentee: Schema.Types.ObjectId,
```
Comment (ethkim) [inline src/server/models/match.model.js]: menteeId here

File: src/server/controllers/match.controller.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+const matchSchema = require('mongoose').model('match');
+import mailer from 'nodemailer';
+
+/*
+ * Methods about mentoring request, accept or reject including E-mail Service
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: In the future, we may need to refactor out the email engine in a separate file so we can use email functions in different part of code. :)

File: src/server/controllers/match.controller.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+// Send mentoring request pushing Email to mentor(receiver)
+function sendRequestEmail(res, mentor, mentee, content) {
+  let transport
+    = mailer.createTransport('smtps://yoda.mentor.lab%40gmail.com:svbootcamp@!@smtp.gmail.com');
+  let mailOptions = {
+    from: YODA_ACCOUNT,
+    to: mentor,
+    replyTo: mentee,
+    subject: EMAIL_SUBJECT,
+    html: EMAIL_HTML + content,
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: Add a TODO to properly format the email body. Right now, we add an HTML header which may be followed by just a string text.

File: src/server/models/match.model.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+let matchSchema = new Schema({
+  mentorId: Schema.Types.ObjectId,
+  menteeId: Schema.Types.ObjectId,
+  requestDate: {
+    type: Date,
+    default: Date.now(),
+  },
+  isAccepted: {
+    type: Number,
```
Comment (ethkim) [inline src/server/models/match.model.js]: Why are these Number instead of Boolean? Same for isCancelled below.

File: src/server/models/match.model.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    default: 0,
+  },
+  acceptDate: Date,
+  isCanceled: {
```
Comment (ethkim) [inline src/server/models/match.model.js]: isCanceled -> isCancelled

File: src/server/models/match.model.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+import mongoose from 'mongoose';
+const Schema = mongoose.Schema;
+
+let matchSchema = new Schema({
+  mentorId: Schema.Types.ObjectId,
```
Comment (ethkim) [inline src/server/models/match.model.js]: Just so I understand how this works: matchSchema here and matchData above don't seem to have the same set of fields. How does this work then? On line 14 of match.controller.js, matchData seem to be instantiating from matchSchema. 

For this to work, we'd have to store things like:

matchData.mentorId = req.mentorId // get the mentorId from request
matchData.menteeId = req.session.user_id // get the current session's user ID.
... etc.

For now, you can just add a TODO and remove lines 65,66, 76, 77 from match.controller.js since those methods aren't complete yet.

**Comment by ethkim** [comment]
Last few comments! After these fixes, the rest of the PR LGTM!

File: src/server/models/match.model.js (hunk 17 above)
Comment (KaylaKwon) [inline src/server/models/match.model.js]: because this is an ObjectId, email address should be called from DB.
But I didn't connect DB yet so hard coded in match.controller.js
I'll add TODO

File: src/server/controllers/match.controller.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+export function requestMentoring(req, res, next) {
+
+  matchData.save((err) => {
+    if (err) {
+      sampleFailResult.errPoint = 'RequestMentoring - Saving MatchData';
+      sampleFailResult.err = err;
+      res.json(sampleFailResult);
+    } else {
+      // Todo:
+      // 1. update mentor and mentee's request db
```
Comment (ifndefdeadmau5) [inline src/server/controllers/match.controller.js]: I Just wondering why 'update' in todo1 didn't capitalized while todo2 did

Comment (ethkim) [inline src/server/controllers/match.controller.js]: Yup, please capitalize the first char in all lines in the comments, and Todo -> TODO

**Comment by ethkim** [comment]
LGTM after the previous comments are reflected. Thanks Kyungah!
